crates/traces/src/lib.rs

```diff
@@ -0,0 +1,253 @@
+use anvil_zksync_types::traces::{
+    CallLog, CallTrace, CallTraceArena, CallTraceNode, DecodedCallEvent, DecodedCallTrace, L2L1Log,
+    L2L1Logs, TraceMemberOrder, KNOWN_ADDRESSES,
+};
+use decode::CallTraceDecoder;
+use writer::TraceWriter;
+use zksync_multivm::interface::{Call, VmExecutionResultAndLogs};
+use zksync_types::H160;
+
+pub mod abi_utils;
+pub mod decode;
+pub mod identifier;
+pub mod writer;
+
+/// Converts a single call into a CallTrace.
+#[inline]
+fn convert_call_to_call_trace(
+    call: &Call,
+    depth: usize,
+    tx_result: VmExecutionResultAndLogs,
+) -> CallTrace {
+    let label = KNOWN_ADDRESSES
+        .get(&call.to)
+        .map(|known| known.name.clone());
+    CallTrace {
+        depth,
+        success: !tx_result.result.is_failed(),
+        caller: call.from,
+        address: call.to,
+        execution_result: tx_result,
+        decoded: DecodedCallTrace {
+            label,
+            ..Default::default()
+        },
+        call: call.clone(),
+    }
+}
+
+/// Builds a call trace arena from the calls and transaction result.
+pub fn build_call_trace_arena(
+    calls: &[Call],
+    tx_result: &VmExecutionResultAndLogs,
+) -> CallTraceArena {
+    let mut arena = CallTraceArena::default();
+
+    // Update the root node's execution result.
+    if let Some(root_node) = arena.arena.get_mut(0) {
+        root_node.trace.execution_result = tx_result.clone();
+    }
+
+    for call in calls {
+        process_call_and_subcalls(call, 0, 0, &mut arena, tx_result);
+    }
+    arena
+}
+
+/// Recursively process a call and its subcalls, adding them to the arena.
+fn process_call_and_subcalls(
+    call: &Call,
+    parent_idx: usize,
+    depth: usize,
+    arena: &mut CallTraceArena,
+    tx_result: &VmExecutionResultAndLogs,
+) {
+    // Collect logs for the current call.
+    let logs_for_call: Vec<CallLog> = tx_result
+        .logs
+        .events
+        .iter()
+        .enumerate()
+        .filter_map(|(i, vm_event)| {
+            if vm_event.address == call.to {
+                Some(CallLog {
+                    raw_log: vm_event.clone(),
+                    decoded: DecodedCallEvent::default(),
+                    position: i as u64,
+                })
+            } else {
+                None
+            }
+        })
+        .collect();
+
+    // Collect user and system L2-L1 logs associated with this call.
+    let l1_l2_logs_for_call: Vec<L2L1Logs> = tx_result
```

Comment:
Here and other places
```suggestion
    let l2_l1_logs_for_call: Vec<L2L1Logs> = tx_result
```